Atru/create meta args reference topics (#1179)
hashicorp/web-unified-docs: content/terraform/v1.13.x/docs/language/meta-arguments/providers.mdx

---
page_title: providers meta-argument reference
description: Learn how the `providers` meta-argument works in Terraform configuration.
---

# `providers` reference

By default, child modules inherit the default provider configurations of their parent module. You can specify an alternate provider configuration in the `module` block using the `providers` argument. The `providers` argument instructs Terraform to use the reference provider configuration to create the module resources.

## Usage

> In a [`module` block](/terraform/language/module/syntax), you can add the optional `providers` meta-argument to specify which [provider configurations](/terraform/language/modules/configuration) from the parent module is available inside the child module.

Author: ```suggestion
In a [`module` block](/terraform/language/block/module), you can add the optional `providers` meta-argument to specify which [provider configurations](/terraform/language/modules/configuration) from the parent module is available inside the child module.
```

refer the specified configurations from the parent module.

In the following example, Terraform uses the default `aws` configuration for AWS resources in the root module where no explicit provider instance is selected, but the example child moduleuses the alternate configuration with the alias `usw2`. As a result, any AWS resources the module defines uses the `us-west-2` region.

```hcl
provider "aws" {
  region = "us-west-1"
}

provider "aws" {
  alias  = "usw2"
  region = "us-west-2"
}

module "example" {
  source    = "./example"
  providers = {
    aws = aws.usw2
  }
}
```

### Default behavior

The `providers` argument is optional when the child module does not declare a [`configuration_aliases`](/terraform/language/modules/develop/providers#provider-aliases-within-modules). When you omit the argument, a child module inherits all of the default provider configurations from its parent module. Default provider configurations are configurations that don't have an `alias` argument.

Specifying a `providers` argument cancels the default behavior. As a result, the child module only has access to the provider configurations you specify.

> 

Reviewer: ```suggestion
```

Extra newline

child module, refer to [Providers Within Modules](/terraform/language/modules/develop/providers).

## Supported constructs

You can use `providers` in the following Terraform configuration blocks:

- [`module` blocks](/terraform/language/block/module)

### Example use cases

The following use cases describe common patterns for the `providers` argument.

### Apply different provider configurations for module resources

In the following example, the `tunnel` module includes resources that support their own provider configuration. As a result, each resource alias maps to a different provider configurations declared in the root module:

```hcl
provider "aws" {
  alias  = "usw1"
  region = "us-west-1"
}

provider "aws" {
  alias  = "usw2"
  region = "us-west-2"
}

module "tunnel" {
  source    = "./tunnel"
  providers = {
    aws.src = aws.usw1
    aws.dst = aws.usw2
  }
}
```

### Modules with alternate provider configurations

In rare cases, a single re-usable module might require multiple configurations
of the same provider. For example, a module that configures connectivity between
networks in two AWS regions is likely to need both a `source` and a `destination`

> region. The following example shows possible root module configuration for this use case:

Author: ```suggestion
region. 

In the following example, the `tunnel` module includes resources that support their own provider configuration. As a result, each resource alias maps to a different provider configurations declared in the root module:
```

> ```hcl

Reviewer: Is this not the same example as above? I like this extra context in this copy of it though :)
